### DISCUSSION TITLE
Link flag unexpectedly detected as `/defaultlib:msvcrt` on debug build, expecting `/defaultlib:msvcrtd`

### DISCUSSION THREAD

**Discussion by Billyzou0741326** [corrosion-rs/corrosion#536] [Q&A] [answered]
Not sure if this is an issue or if this is expected behavior. Thought that discussion would be a good place to start.

My setup uses `corrosion@v0.5`, `cmake@3.29.2` with `-DRust_TOOLCHAIN=stable-x86_64-pc-windows-msvc` and MSVC 2019 C and CXX compiler.

When I build with `Debug` config, I'm seeing this link time warning:

```
(Link target) ->
  LINK : warning LNK4098: defaultlib 'msvcrt' conflicts with use of other libs; use /NODEFAULTLIB:library [C:\Users\ktr\Documents\Projects\c\single001\build\test\c001.vcxproj]
```

After digging into the build configs generated by `-DCMAKE_GENERATOR:STRING=Visual Studio 16 2019` and `-DCMAKE_GENERATOR:STRING=Ninja` generators respectively, I found the link flags `/defaultlib:msvcrt` being included:

```
  // c001.vcxproj
  <ItemDefinitionGroup Condition="'$(Configuration)|$(Platform)'=='Debug|x64'">
    ...
    <Link>
      <AdditionalDependencies>c001_rust.lib;kernel32.lib;advapi32.lib;ntdll.lib;userenv.lib;ws2_32.lib;kernel32.lib;advapi32.lib;ntdll.lib;userenv.lib;ws2_32.lib;msvcrtd.lib;kernel32.lib;user32.lib;gdi32.lib;winspool.lib;shell32.lib;ole32.lib;oleaut32.lib;uuid.lib;comdlg32.lib;advapi32.lib</AdditionalDependencies>
      ...
      <AdditionalOptions>%(AdditionalOptions) /machine:x64 /defaultlib:msvcrt</AdditionalOptions>
```

```
  // build.ninja
  LINK_FLAGS = /machine:x64 /debug /INCREMENTAL /subsystem:console  /defaultlib:msvcrt
  LINK_LIBRARIES = c001_rust.lib  kernel32.lib  advapi32.lib  ntdll.lib  userenv.lib  ws2_32.lib  kernel32.lib  advapi32.lib  ntdll.lib  userenv.lib  ws2_32.lib  msvcrtd.lib  kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib
```

Note the `/defaultlib:msvcrt` are present in **`<AdditionalOptions>`** and **`LINK_FLAGS`**. For the sake of testing, I manually changed `/defaultlib:msvcrt` to `/defaultlib:msvcrtd`, and the warning went away.

In contrast, `msvcrtd.lib` are both correctly detected in **`<AdditioalDependencies>`** and **`LINK_LIBRARIES`**, thanks to the logic here:

https://github.com/corrosion-rs/corrosion/blob/64289b1d79d6d19cd2e241db515381a086bb8407/cmake/FindRust.cmake#L192

where the `libs_list` correctly substitudes `msvcrt` with `msvcrtd` on Debug builds.

However, the `flags_list` contains `/defaultlib:msvcrt`, and is not getting substituded:

https://github.com/corrosion-rs/corrosion/blob/64289b1d79d6d19cd2e241db515381a086bb8407/cmake/FindRust.cmake#L181-L182

Not sure if this is the intended behavior? I think it would make sense to do similar substitutions for the link flags. Though I'm not familiar with cmake scripting at all, something to note would be that `/defaultlib:xxx` and `/DEFAULTLIB:xxx` are both valid as link flags are case-insensitive according to [msvc docs](https://learn.microsoft.com/en-us/cpp/build/reference/linker-options?view=msvc-170)

**Comment by jschwe** [comment, marked as answer]
Thanks for the detailed report!

> Not sure if this is the intended behavior? I think it would make sense to do similar substitutions for the link flags.

No, that's not intended behavior, we should replace it too. We only need to replace the variant that `rustc` uses, which is the lowercase `/defaultlib:xxx`. 
This is a quick fix, I can do it later today.

**Reply by Billyzou0741326** [reply]
I took a stab at this https://github.com/Billyzou0741326/corrosion/commit/101507bc732670031be3c5bce41c4c96eefc72ae and observed that it builds fine when I point my project to that commit hash.

If that looks acceptable, let me know if I should open a PR (either against `master` or `stable/v0.5`, or both?)

**Reply by jschwe** [reply]
I initially had something similar in mind, but the Generator expression `"$<$<CONFIG:Debug>:/defaultlib:msvcrtd>"` actually expands to nothing for all configurations except "Debug" (including the most common build configuration `""` (empty string). That then lead to https://github.com/corrosion-rs/corrosion/pull/537
